### PR TITLE
feat: add pipefy_webhook resource 

### PR DESCRIPTION
**Pull request by looslidias**
## Summary

Adds `pipefy_webhook`, an importable resource that manages pipe webhooks through `createWebhook` / `updateWebhook` / `deleteWebhook`. Closes #44.

Webhooks are the kind of integration config teams most want in code instead of clicked into the UI. Until now a config could only register one by shelling out to the API.

## Schema

- `pipe_id` (required, RequiresReplace): the update mutation has no `pipe_id`, so moving a webhook to another pipe replaces it.
- `url` (required): validated with a reusable `URL()` validator that accepts any absolute `http` or `https` URL. `http` is allowed for local endpoints; the HTTPS requirement is left to the backend at apply time.
- `actions` (required, list of strings): free-text, since the supported triggers change on the backend. The description points at the API docs rather than hardcoding an enum.
- `name` (required): the create input mandates a non-null name.
- `headers` (optional, sensitive) and `filters` (optional): JSON, typed with `jsontypes.Normalized`.

## Design notes

- `headers` and `filters` map to two different GraphQL scalars. `headers` is the `Json` scalar, which expects a JSON document encoded as a string, so the raw value is sent through unchanged. `filters` is the `JSON` scalar, which expects an object, so the value is unmarshaled before it is sent. Sending an object for `headers` is rejected by the API; this was confirmed live and is guarded by a test.
- Both use `jsontypes.Normalized` instead of a hand-rolled validator. It validates JSON at plan time with no custom code and applies semantic equality, so reformatting the JSON (whitespace, key order) does not produce a spurious diff.
- `filters` is refreshed from the API so a filter edited outside Terraform is detected as drift. Read keeps the API's raw JSON rather than unmarshaling and re-marshaling, so numeric IDs above 2^53 never lose precision through float64. `headers` is not read back because it is sensitive: the configured value is authoritative and re-sent on every apply. As a result, after an import the first plan re-sends `headers` only; `filters` imports cleanly.
- Update sends an explicit `null` for `headers` and `filters` when they are removed from config, so the remote values are cleared instead of lingering. This pairs with refreshing `filters`: without it, a removed filter would refresh back into a permanent diff.
- Plan-time validation is limited to static and format checks: the `url` scheme and the JSON validity of `headers` / `filters`. Business rules such as "filters allow only one action" are left to the API and surfaced from its response at apply time, per the provider's plan-time vs apply-time guidance.
- There is no root `webhook(id)` query. Read goes through `pipe(pipe_id).webhooks` and finds by webhook id, so import takes the composite id `pipe_id/webhook_id`. The field selection lives in a `webhookgql` package with no client dependency, matching `labelgql`.

## Testing

- httptest-based tests: full CRUD plus import (including the malformed-id error); the filters happy path (sent as a JSON object, refreshed without a spurious diff, survives a value change); filters drift detection; clear-on-remove (asserts the update payload carries `headers: null` / `filters: null`); and the plan-time validations (URL and JSON on `headers` / `filters`). An in-package unit test covers `normalizeFilters`, including the 2^53 precision case.
- Verified live against a local Pipefy instance: `headers` must be sent as a JSON string (an object is rejected) and is returned verbatim; `filters` round-trips as an object against real phase ids and an out-of-band change is detected as drift; removing `headers` / `filters` clears them on the API; an empty diff holds after apply even with reformatted headers JSON; and import round-trips.

## Scope

Pipe webhooks only. The API also accepts `table_id`, `email`, and `identityPropagationEnabled`; those are left out for now and can follow up if needed.

### CODE REVIEW

File: internal/provider/resources/resource_webhook.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	if value.IsNull() || value.ValueString() == "" {
+		return true
+	}
+	var v any
+	diags.Append(value.Unmarshal(&v)...)
+	if diags.HasError() {
+		return false
+	}
+	input["filters"] = v
+	return true
```
Comment (adriannoes) [inline internal/provider/resources/resource_webhook.go:299]: On Update, when `headers` or `filters` go from set → null/empty in the plan, these helpers omit them from the mutation input. For GraphQL partial updates that means the remote webhook keeps the previous values — including any Authorization secret — while Terraform state looks cleared, and Read will never reveal the drift (selection is only `id name url actions`).

Please send an explicit `null` (or whatever clearing sentinel the API accepts) when the attribute is being removed, and cover it with a unit test that asserts the update variables contain `"headers": null` / `"filters": null`. You'll know it's fixed when removing those attrs from config and applying leaves the remote webhook without the old headers/filters.

File: internal/provider/resources/resource_webhook.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+				Optional:    true,
+				Sensitive:   true,
+				CustomType:  jsontypes.NormalizedType{},
+				Description: "Custom HTTP headers sent with the webhook, as a JSON object string (e.g. \"{\\\"Authorization\\\":\\\"Bearer ...\\\"}\"). Not read back from the API, so changes to it are only reconciled on the next apply.",
+			},
+			"filters": schema.StringAttribute{
+				Optional:    true,
+				CustomType:  jsontypes.NormalizedType{},
+				Description: "Filters that restrict when the webhook fires, as a JSON string. Only one action can be configured when filters are used. See https://developers.pipefy.com/reference for the supported keys per action.",
+			},
```
Comment (adriannoes) [inline internal/provider/resources/resource_webhook.go:76]: The `headers` description says changes are reconciled on the next apply, and `filters` doesn't mention write-only behavior at all even though Read treats both the same way. That steers users toward believing `terraform plan` / removal will reflect remote state.

Please align both attribute descriptions (and the generated docs) with the real contract: write-only, not refreshed on Read, and clear-on-remove only once the Update path actually clears them (or call out that clear isn't supported yet).

**Review by adriannoes — CHANGES_REQUESTED** [review]
Solid webhook resource, @looslidias,  CRUD, import, Json-vs-JSON handling, plan-time validators, examples, and CI all look good. Local checkout at `2038782` passed `make build`, `make lint`, `make test`, and `make generate` (no docs drift).

**Required before merge**

- **Clearing `headers` / `filters` on Update** — omitting a null attribute does not clear the remote webhook, and Read never refreshes those fields, so secrets and filter rules can survive a config removal invisibly. Please send an explicit clear (`null` or a verified API sentinel) when the plan removes them, plus a unit test asserting the update payload.
- **Schema/docs for write-only attrs** — `headers` currently reads like any change (including removal) is reconciled on apply; `filters` has no write-only note at all. Align both descriptions with the real (post-fix) contract and regenerate docs.

**Also noted**

- `deleteWebhook` ignores `success: false` (same pattern as most siblings) — worth checking before dropping state.
- Serialization type assertions only cover create; Recording headers/filters types on `updateWebhook` too would lock the PR’s main guardrail for header rotation.

File: internal/provider/resources/resource_webhook.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+	var data WebhookModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	filtersSet := !data.Filters.IsNull() && !data.Filters.IsUnknown() && data.Filters.ValueString() != ""
+	if !filtersSet || data.Actions.IsNull() || data.Actions.IsUnknown() {
+		return
+	}
+	if len(data.Actions.Elements()) != 1 {
```
Comment (Danielmoraisg) [inline internal/provider/resources/resource_webhook.go]: This enforces a backend business rule (exactly one action when filters are set) at plan time. The repo's own guidance (CLAUDE.md, "Plan-Time vs. Apply-Time") leans toward deferring dynamic business rules to the API and reserving plan-time for static or format checks.

File: internal/provider/resources/resource_webhook.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+	Filters jsontypes.Normalized `tfsdk:"filters"`
+}
+
+func (r *WebhookResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_webhook"
+}
+
+func (r *WebhookResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Webhook resource",
```
Comment (Danielmoraisg) [inline internal/provider/resources/resource_webhook.go]: Nit: this string becomes the top line of the generated doc (`docs/resources/webhook.md` renders "Webhook resource" twice). A one-line description of what the resource manages reads better, matching `pipefy_pipe_relation`'s "Connects a parent pipe to a child pipe..."

File: internal/provider/resources/resource_webhook.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+	data.Name = types.StringValue(w.Name)
+	data.Url = types.StringValue(w.Url)
+	actions, d := types.ListValueFrom(ctx, types.StringType, w.Actions)
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	data.Actions = actions
+	// headers and filters are intentionally left as-is: headers are sensitive
+	// and filters are kept as the user's JSON string to avoid diffs.
```
Comment (Danielmoraisg) [inline internal/provider/resources/resource_webhook.go]: suggestion: `filters` is a good candidate for reading back from the API, unlike `headers`. It isn't sensitive, and it's typed `jsontypes.Normalized`, which compares JSON semantically.

I verified against the live API that reading it back does not produce spurious diffs. I created webhooks with single-element, multi-element, reversed-order, and multi-key numeric filters, captured what `pipe.webhooks` returns, and ran each (config, API-returned) pair through the actual `Normalized.StringSemanticEquals`, both storing the raw bytes and re-marshaling through a Go value. All four are semantically equal, no diff. The API preserves array order, object keys, and numeric types; it does not reorder or canonicalize.

So reading `filters` back would give real drift detection that the current code silently skips (a filter edited in the UI is invisible to `plan`).

Two implementation notes if you do this:
- Capture the API's raw JSON (`json.RawMessage`) rather than unmarshal then re-marshal, so numeric IDs above 2^53 can never lose precision through float64. Current Pipefy IDs are ~9 digits, well under that, but raw bytes sidesteps it entirely.
- Pair it with the clear-on-remove fix. If `filters` is read back but Update still omits a removed `filters`, the read refreshes the stale value from the API and you get a permanent diff instead of silent drift. The two changes are complementary.

File: internal/provider/resources/resource_webhook.go (hunk 1 above)
Comment (looslidias) [reply internal/provider/resources/resource_webhook.go:299]: Thanks for noticing it.

Update now sends an explicit null for headers and filters when they are removed from config, via updateHeadersInput /updateFiltersInput, so the remote values are cleared instead of lingering.

File: internal/provider/resources/resource_webhook.go (hunk 2 above)
Comment (looslidias) [reply internal/provider/resources/resource_webhook.go:76]: Let's improve drifting detection, right?

One deliberate divergence from "write-only for both": following the discussion about reading filters back, filters is now refreshed from the API so drift is detected, while headers stays write-only since it is sensitive.

File: internal/provider/resources/resource_webhook.go (hunk 3 above)
Comment (looslidias) [reply internal/provider/resources/resource_webhook.go]: Thanks for noticing it!

That's definitely a dynamic business rule that should be deferred to Pipefy's API, as we have already stated. Double checked and, gladly, that indeed returns "Validation failed: Webhooks with filters may only contain one action" error.

File: internal/provider/resources/resource_webhook.go (hunk 4 above)
Comment (looslidias) [reply internal/provider/resources/resource_webhook.go]: Changed to "Sends an HTTP request to a URL when the selected events occur on a pipe."

File: internal/provider/resources/resource_webhook.go (hunk 5 above)
Comment (looslidias) [reply internal/provider/resources/resource_webhook.go]: Thanks for the detailed verification.

Read now refreshes filters via normalizeFilters, which keeps the API's raw JSON (json.RawMessage) instead of unmarshaling then re-marshaling, so numeric IDs above 2^53 never round-trip through float64; an empty object or null maps to a null attribute so it matches an unset config. Paired with the clear-on-remove fix, so a removed filter does not refresh back into a permanent diff. Covered by TestUnit_WebhookResource_FiltersDrift for out-of-band drift and TestNormalizeFilters_LargeIntPrecision for the 2^53 case.

**Review by adriannoes — APPROVED** [review]
Follow-up looks good — approving.

Clear-on-remove for `headers`/`filters` plus refreshing `filters` on Read is the right pairing: without both, you'd either leave remote secrets/filters behind or get a permanent drift diff. Docs and tests match that contract. Nice work.

**Review by Danielmoraisg — APPROVED** [review]
